ext/nio4r/monitor.c

```diff
@@ -60,13 +60,18 @@ void Init_NIO_Monitor()
 static VALUE NIO_Monitor_allocate(VALUE klass)
 {
     struct NIO_Monitor *monitor = (struct NIO_Monitor *)xmalloc(sizeof(struct NIO_Monitor));
-
+    
+    if(!monitor)
+        return Qnil;
+    
+    *monitor = (struct NIO_Monitor){.self = Qnil};
     return Data_Wrap_Struct(klass, NIO_Monitor_mark, NIO_Monitor_free, monitor);
 }
 
 static void NIO_Monitor_mark(struct NIO_Monitor *monitor)
 {
-    return rb_gc_mark(monitor->self);
+    if(monitor && monitor->self != Qnil)
+        rb_gc_mark(monitor->self);
```

Comment:
I think this is redundant if `monitor->self` is initialized to `Qnil`.

Reply:
Which part is redundant? Testing that `monitor` in non-`NULL` or testing that `monitor->self` is non-`nil`?

I would have agreed on both, but this line was the only difference that caused Ruby to crash (see #244)... so it must be one of these values.

Reply:
Or maybe it's something else altogether and I'm shooting arrows at the sun.

Reply:
If the core problem is really that `NIO_Monitor_mark` is getting called after `NIO_Monitor_allocate` but before `NIO_Monitor_initialize`, then initializing the struct in `NIO_Monitor_allocate` should be sufficient to address the problem.

AFAIK calling `rb_gc_mark` on `Qnil` should be harmless, since it's a valid object.

I don't think there are any circumstances under which Ruby would invoke functions passed to `Data_Wrap_Struct` with a NULL pointer, but if you'd really like to check for that, I think an `assert()` would make more sense.

Reply:
I thought that this was the situation.

However, the issue in #244 logs:

```text
[BUG] try to mark T_NONE object
ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]
```

Both of this error message suggests that `rb_gc_mark(0)` might blow up.

Reply:
That's not `Qnil` then, and something else is wrong.

`T_NONE` is equal to 0, and is not a valid Ruby object.
`Qnil` is either 4 or 8, depending on whether or not `USE_FLONUM` is defined:

https://github.com/ruby/ruby/blob/9e41a75/include/ruby/internal/special_consts.h#L79-L89

So for some reason `monitor->self` is being set to `NULL` rather than `Qnil`

Reply:
Fair point.

I can either update the `if` to `(monitor && monitor->self)` or to `(monitor->self)`, as you desire. Let me know.

Side-note:  AFAIK, historically either `Qfalse` or `Qnil` were `==0`... but maybe that was just a rumor.

Reply:
I just noticed the link you added - `Qfalse` is `== 0`, which means that `rb_gc_mark(0)` really shouldn't explode...

I'm not sure what's going on there. It might be a "junk" data issue due to the fact that `monitor` wasn't initialized, so maybe that's all we really need.

Again, let me know what you prefer, I can simply remove the `if` and leave the memory initialization in place, but Idon't see a reason not to test.

Reply:
`Qnil` was always 4 in earlier versions of Ruby:

https://blog.bigbinary.com/2008/06/23/why-the-id-of-nil-is-4-in-ruby.html

If `rb_gc_mark(Qnil)` succeeds, I don't think we should do any kind of gating on when it's called. `monitor->self` should always point to a valid Ruby object.

Instead, I'd suggest adding something like: `assert!(monitor->self != NULL)`. If it is `NULL` (which it appears to be), we have other problems.

Reply:
`monitor->self == NULL` is the same as `monitor->self != Qfalse`, which means that no gating is necessary.

If you don't like the `if` statements, I can reduce the whole thing to memory initialization in a new PR and leave it that...?

Reply:
Memory initialization and `assert`s sounds good to me.

The reason for using `assert` over `if` is the conditions in question should not occur during normal operation (or if they do, we should figure out why and document it). That is to say, they're bugs and shouldn't be happening.

Reply:
Done. I guess this means we resolved the issues in the review?

Reply:
Yep, looks good (although you might want to update the PR title to reflect the changes)

Reply:
Done (renamed).